### PR TITLE
Replay Player Fixes

### LINKED ISSUE #341: Replay misses some/many notes after pressing windows key
**Describe the bug**
When you watch a replay and you press the windows key, the replay misses some/many notes + it dies when the HP is too low. (Depends on how hard and fast the map is)

**To Reproduce**
1. Go to a map which has a replay on it
2. Click on 'Watch Replay'
3. Press the 'Windows-Key'

**Expected behavior**
I expected nothing to happen. Neither missing a note, nor failing the entire map.

**Screenshot**
After pressing the windows key
0screenshot

**Desktop (please complete the following information):**
 -OS: Windows 10

**Additional context**
I don't know if it's supposed to be like that and I don't know if that's only an issue for me.

### PR DESCRIPTION
**Pull request by Swan**
Fixes #341 

Makes scoring for replays accurate & fixes autoplay failing when lagging. This solution preprocesses replays with [VirtualReplayPlayer](https://github.com/Quaver/Quaver.API/blob/master/Quaver.API/Replays/Virtual/VirtualReplayPlayer.cs) and uses those calculated judgements for scoring.

Still needs a bit of work, as there are still some *visual* bugs (w/ judgements & LNs) when completely freezing the window, but this is a decent start.

**Edit**

Said visual bugs have been fixed. Only thing remaining is making sure the replay player catches up for missed frames

### CODE REVIEW

File: Quaver.Shared/Screens/Gameplay/Rulesets/Input/ReplayInputManagerKeys.cs [hunk 1]
```diff
@@ -46,6 +50,16 @@ public class ReplayInputManagerKeys
         /// </summary>
         internal List<bool> UniqueReleases { get; } = new List<bool>();
 
+        /// <summary>
+        ///     Virtually plays replay frames
+        /// </summary>
+        public VirtualReplayPlayer VirtualPLayer { get; }
```
Comment (YaLTeR) [inline Quaver.Shared/Screens/Gameplay/Rulesets/Input/ReplayInputManagerKeys.cs]: PLayer

File: Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs [hunk 2]
```diff
@@ -97,8 +98,6 @@ public ResultHitDifferenceGraph(ScalableVector2 size, ResultScreen screen)
 
             // Make some fake hits for debugging.
             // CreateFakeHitStats();
-
-            // Draw the dots if there are any.
```
Comment (YaLTeR) [inline Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs]: How did the comment end up in a different place

File: Quaver.Shared/Screens/Gameplay/Rulesets/Input/ReplayInputManagerKeys.cs [hunk 3]
```diff
@@ -103,6 +122,37 @@ internal void HandleInput()
...
+                    ((ScoreProcessorKeys)Screen.Ruleset.ScoreProcessor).CalculateScore(judgement);
+
+                    // Update Scoreboard
+                    var view = (GameplayScreenView) Screen.View;
+                    view.UpdateScoreAndAccuracyDisplays();
+
+                    var playfield = (GameplayPlayfieldKeys)Screen.Ruleset.Playfield;
+                    playfield.Stage.ComboDisplay.MakeVisible();
+
+                    if (judgement != Judgement.Miss)
```
Comment (YaLTeR) [inline Quaver.Shared/Screens/Gameplay/Rulesets/Input/ReplayInputManagerKeys.cs]: What about misses?

Comment (Swan) [reply Quaver.Shared/Screens/Gameplay/Rulesets/Input/ReplayInputManagerKeys.cs]: Misses don't show in the hit error

File: Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs [hunk 4]
```diff
@@ -162,6 +163,12 @@ private void CreateFakeHitStats()
         /// <exception cref="ArgumentOutOfRangeException"></exception>
         private static ScoreProcessor GetScoreProcessor(ResultScreen screen)
         {
+            if (screen.Gameplay != null && screen.Gameplay.InReplayMode)
+            {
+                var im = screen.Gameplay.Ruleset.InputManager as KeysInputManager;
+                return im?.ReplayInputManager.VirtualPlayer.ScoreProcessor;
+            }
```
Comment (YaLTeR) [inline Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs]: Doesn't the case at the bottom cover replays already?

Comment (Swan) [reply Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs]: Yes, but it also reprocesses the replay which isn't necessary anymore.

Comment (YaLTeR) [reply Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs]: Could you add a comment about that?

Comment (Swan) [reply Quaver.Shared/Screens/Result/UI/ResultHitDifferenceGraph.cs]: done

**Pull request merged**